README


Voici quelques indications pour TESTER l'application

=====================================================
La documentation pour la mise en production est sur
file://usr/share/doc/auf-django-users/html/index.html
=====================================================


A. INITIALISATION
-----------------

1. Ne PAS toucher au fichier de configuration /etc/auf-django-users/conf.py
   Ce fichier de configuration par défaut est fait pour les tests.

       # extrait de /etc/auf-django-users/conf.py (début du fichier)
       DATABASE_ENGINE = 'sqlite3'
       DATABASE_NAME = '/tmp/base.sqlite3'

   Par défaut, la base est en sqlite3 dans /tmp/base.sqlite3
   Note : si vous aviez déjà testé une ancienne version du projet,
   cette base peut ne plus être compatible ! Dans ce cas, effacez
   d'abord l'ancienne base (rm /tmp/base.sqlite3)

2. Initialiser cette base de données de test :

     $ auf-django-users-manage.py syncdb

   Note : un compte administrateur sera créé avec le login et le mot de passe
   que vous indiquerez.

3. Insérer un utilisateur et un groupe initial dans la base :

     $ auf-django-users-manage.py loaddata utilisateur_test


B. LANCEMENT DE L'INTERFACE WEB (en mode "test")
------------------------------------------------

1. Lancer le serveur de test :

     $ auf-django-users-manage.py runserver

2. Aller sur http://127.0.0.1:8000/admin/
   Connectez-vous avec le compte administrateur (créé lors de l'initialisation).

   En tant qu'administrateur vous voyez absolument tout le système Django.
   Vous verrez TOUTES les applications, seules CERTAINES nous intéressent:

     Auth                          --> système d'authentification web
      Groupes                          de Django en interne : cette appli
      Utilisateurs                     ** ne concerne pas notre test **

     Logs                          --> notre suivi de renouvellement des
      Modifications d'expiration       abonnements (i.e. expirations)

     Mail                          --> Gestion des comptes de messagerie
      Comptes de messagerie            synchro avec nss (cf ci dessous)

     Nss                           --> ** NOTRE APPLICATION MERE ! **
      Groupes systèmes                 C'est par ici qu'il faut cliquer
      Utilisateurs systèmes

   Finalement, seule la table "Utilisateurs système" nous intéresse. Cliquez
   dessus... et jouez !

C. OBSERVATION DE LA BASE DE DONNEES
------------------------------------

Les données sont stockées dans /tmp/base.sqlite3. Pour regarder ce
qu'elle contient, utilisez sqlite3. Par exemple :

 * les utilisateurs :
    $ sqlite3 -line /tmp/base.sqlite3 "select * from users;"
 * les groupes :
    $ sqlite3 -line /tmp/base.sqlite3 "select * from groups;"
 * les associations de groupes secondaires :
    $ sqlite3 -line /tmp/base.sqlite3 "select * from grouplist;"
 * suivi des abonnements :
    $ sqlite3 -line /tmp/base.sqlite3 "select * from log;"

Si vous activez la contrib.mail (voir fin de conf.py) :
 * les utilisateurs messagerie :
    $ sqlite3 -line /tmp/base.sqlite3 "select * from mail_users;"




debian/changelog
auf-django-users (0.5.14) unstable; urgency=low

  * aufusers.__authors__ et __version__ (à mettre à jour à chaque
    fois, en attendant de trouver une ruse)
  * debian/postinst: creation d'un groupe auf-d-u + modifications
    des droits par défaut du fichier /etc/a-d-u/conf.py (640 root:auf-d-u)
    + ajout de www-data dans le groupe auf-d-u

 -- Thomas NOEL <thomas.noel@auf.org>  Fri, 31 Jul 2009 16:01:42 +0000

auf-django-users (0.5.13) unstable; urgency=low

  * nss/models.py: un print qui trainait...
  * nss/models.py: save() lance une exception si on tente de
    l'appeler avec une source non LOCAL (au lieu d'un simple return
    qui donnait l'impression que tout allait bien... grmbll)
  * examples/ca2django.py: exemple de synchro base AUF vers auf-django

 -- Thomas NOEL <thomas.noel@auf.org>  Wed, 29 Jul 2009 12:33:30 +0000

auf-django-users (0.5.12) unstable; urgency=low

  * contrib/log/models.py: affichage des logs plus "humain"

 -- Thomas NOEL <thomas.noel@auf.org>  Wed, 22 Jul 2009 17:22:02 +0000

auf-django-users (0.5.11) unstable; urgency=low

  * contrib/log/models.py: gestion des details de log dans l'ancien
    format. On peut récupérer les infos via l'attribut (property) details_dict
    qui sera toujours un dict, quel que soit le format de details (JSON ou
    ancien format). C'est un peu bidouillé, mais je paye de prix d'une erreur
    de conception initiale... Et si vous êtes pas content, c'est pareil.

 -- Thomas NOEL <thomas.noel@auf.org>  Tue, 21 Jul 2009 15:32:08 +0000

auf-django-users (0.5.10) unstable; urgency=low

  * nss/models.py: verification formats username, gecos ; raise
    d'exception en cas de tentative de modif de uid ou source
  * nss/admin.py: affichage de l'uid dans les données systèmes
  * contrib/log: enregistrement des details en json
  * debian/control: dependance sur python-simplejson

 -- Thomas NOEL <thomas.noel@auf.org>  Tue, 21 Jul 2009 11:20:16 +0000

auf-django-users (0.5.9) unstable; urgency=low

  * nss/admin.py: clean_gecos et clean_username pour vérification
    des données reçues par le formulaire (uniquement des lettres
    ASCII, chiffres, etc.)
  * urls.py: possibilité de configurer les URLs dans conf.py (petit
    exemple fourni)

 -- Thomas NOEL <thomas.noel@auf.org>  Mon, 20 Jul 2009 14:31:02 +0000

auf-django-users (0.5.8) unstable; urgency=low

  * doc/: description du système de gestion des droits Django,
    et ajout d'annexes concernant auf-poste-client et auf-mkhomedir

 -- Thomas NOEL <thomas.noel@auf.org>  Tue, 09 Jun 2009 10:48:17 +0000

auf-django-users (0.5.7) unstable; urgency=low

  * nss/models.py: import plus générique des données de configuration,
    qui permet de ne pas planter quand quelque chose est oublié dans
    conf.py

 -- Thomas NOEL <thomas.noel@auf.org>  Mon, 08 Jun 2009 11:20:46 +0000

auf-django-users (0.5.6) unstable; urgency=low

  * conf.py: ajout d'un MIN_GID. L'application plante s'il n'existe pas, c'est
    voulu et pas bien méchant, non ?
  * nss/models.py: Group.save() calcule automatiquement un GID
  * nss/fixtures: scission en deux fichiers json group_users et user_test

 -- Thomas NOEL <thomas.noel@auf.org>  Mon, 08 Jun 2009 08:58:47 +0000

auf-django-users (0.5.5) unstable; urgency=low

  * nss/models.py: les champs sont ordonnés pour donner une table SQL
    la plus proche possible des champs passwd et shadow.
  * doc/mysql.rst: description complète des tables
  * nss/models.py Group: ajout d'un attribut Group.users_secgroup (autre façon
    de voir la liaison entre un utilisateur et ses groupes secondaires) et
    d'une méthode users() qui renvoie la liste des utilisateurs d'un group
    (que le groupe leur soit primaire ou secondaire)
  * nss/models.py User: l'attribut groups est renommé secgroups. Une méthode
    groups() renvoie une liste de tous les groupes de l'utilisateur.

 -- Thomas NOEL <thomas.noel@auf.org>  Wed, 03 Jun 2009 17:47:51 +0000

auf-django-users (0.5.4) unstable; urgency=low

  * validation de la modif précédente : ça marche. Ajout d'un README.0.5.4 qui
    explique la procédure d'upgrade manuelle
  * ajout de "(NSS)" dans les verbose_name de User et Group, pour tenter
    d'éviter la confusion avec les user et group de Django

 -- Thomas NOEL <thomas.noel@auf.org>  Wed, 03 Jun 2009 10:47:39 +0000

auf-django-users (0.5.3alpha) unstable; urgency=low

  * 0.5.3alpha, a.k.a. "ici vol DC 132" : version ALPHA à tester !
  * oups oups oups... et aïe aïe aïe : modification du modèle donc 
    du schéma de la base SQL :( Pince à épiler à sortir pour ceux qui sont
    en prod (avec mes plus plates excuses) :
      - username devient la clé de User
      - uid devient un champ integer unique
      - grouplist : association username-gid, et non plus uid-gid
  * Merci à Shafeek pour son oeil de lynx sur ce coup là !!

 -- Thomas NOEL <thomas.noel@auf.org>  Wed, 03 Jun 2009 10:15:10 +0000

auf-django-users (0.5.2) unstable; urgency=low

  * debian/contrib: ajout d'un recommends python-mysqldb

 -- Thomas NOEL <thomas.noel@auf.org>  Mon, 18 May 2009 09:04:32 +0000

auf-django-users (0.5.1) unstable; urgency=low

  * documentation /usr/share/doc/auf-django-users/html mieux
    structurée : on sépare clairement la description de aufusers.nss
    (le noyau) de cette des contribs (log et mail)

 -- Thomas NOEL <thomas.noel@auf.org>  Thu, 14 May 2009 18:03:59 +0000

auf-django-users (0.5) unstable; urgency=low

  * 0.5. Voilà. Fin du début. Ouf.
  * un petit schéma dans la doc
  * logs plus robustes (on ne loggue que dans le post_save, le pre_save
    ne sert qu'à retenir les modifications)
  * logs en couleur et en odorama

 -- Thomas NOEL <thomas.noel@auf.org>  Thu, 14 May 2009 12:43:46 +0000

auf-django-users (0.4.9999) unstable; urgency=low

  * contrib.log : système de log de toutes les modifs des comptes NSS
  * logs devient log_expire et est désactivée par défaut, la
    table associée s'appelle aussi log_expire (et non plus expire_log)

nss/models.py
# -*- encoding: utf-8 -*-
"""

Les classes **User**, **Group** et **GroupList** définies dans ce modèle sont
une représentation des tables ``users``, ``groups`` et ``grouplist`` d'une base
de données suivant le format de ``libnss-mysql-bg``.

Il s'agit de la représentation classique des utilisateurs et groupes sous Unix
pour le système **nss** (Name Service Switch).

"""

from django.db import models
from aufusers.lib.expire import ExpireField
from aufusers.lib.utils import password_crypt
import datetime


# configurations par défault (peuvent être redéfinies dans le conf.py)

# Shells de connexion acceptés , le premier est la valeur par défaut
SHELLS = (
    ('/bin/false', 'Désactivé (/bin/false)'),
    ('/bin/sh', 'Par défaut (/bin/sh)'),
)

# Valeur du "home" par défaut dans le formulaire. Si cette valeur n'est pas
# modifiée lors de la création d'un utilisateur, alors son homedir sera calculé
# par la formule suivante : HOMEBASE + '/' + username
HOME_BASE = "/home"

# Renvoie le homedir d'un utilisateur
def homedir(user):
    return "%s/%s" % (HOME_BASE, user.username)

# UID minimal pour un utilisateur
MIN_UID = 10000
# GID minimal pour un groupe
MIN_GID = 10000
# GID par défaut pour un utilisateur ("users")
DEFAULT_GID = 100

# Renvoie la date d'expiration par défaut
def default_expire(): 
    # retourne la date de fin du mois actuel (j'ai pas mieux comme algo, j'ai
    # peu de neurone) : on se place au debut du mois actuel, on va 32 plus
    # tard, on prend le premier jour du mois obtenu... et on recule d'un jour.
    # Ouf.
    return ( datetime.date.today().replace(day=1) + \
             datetime.timedelta(32) ).replace(day=1) - datetime.timedelta(1)

# import de la configuration : on peut y écraser toutes les définitions
# précédentes
from aufusers.conf import *


class User(models.Model):
    """
        Représentation d'un utilisateur au sens Unix (nss). Voici les champs disponibles :

         * ``username`` : nom de l'utilisateur, uniquement composé de lettres minuscules et de chiffres. C'est la *clé primaire*
         * ``password`` : mot de passe de l'utilisateur, crypté
         * ``uid`` : numéro identifiant l'utilisateur
         * ``gid`` : objet Group auquel appartient l'utilisateur
         * ``secgroups`` : liste des groupes secondaires auquel appartient l'utilisateur (c'est un *ManyRelatedManager*)
         * ``gecos`` : nom et prénom, sous la forme Prenom NOM (sans aucun accent)
         * ``homedir`` : répertoire personnel ($HOME)
         * ``shell`` : interpréteur de commande
         * ``lstchg`` : date du dernier changement du mot de passe (*datetime.date*)
         * ``min`` : nombre de jours à attendre avant de pouvoir changer le mot de passe
         * ``warn`` : nombre de jours d'avertissement avant la fin mot de passe
         * ``max`` : nombre de jours après lesquels le mot de passe doit être changé
         * ``inact`` : nombre de jours après la fin de validité provoquant la désactivation du compte
         * ``expire`` : date d'expiration du compte. Il s'agit d'un objet *datetime.date*
         * ``flag`` : champ réservé
         * ``source`` : source des données de cet utilisateur (== 'LOCAL' si l'utilisateur est géré par l'application)
         * ``creation`` : date de création de l'utilisateur dans la base de donnée
         * ``modification`` : date de la dernière modification des données de l'utilisateur

    """

    username = models.CharField("nom d'utilisateur", max_length=128, primary_key=True,
            help_text="Indiquez un identifiant de connexion, uniquement composé de lettres minuscules et de chiffres")
    password = models.CharField("mot de passe (crypté)", max_length=64, default='x')
    uid = models.IntegerField("userID", unique=True) 
    gid = models.ForeignKey("Group", db_column= "gid", default=DEFAULT_GID,
            help_text="Ne modifiez que si vous savez exactement ce que vous faîtes !")
    secgroups = models.ManyToManyField("Group", related_name='groups', through='GroupList', blank=True)
    gecos = models.CharField("informations GECOS", max_length=128, blank=True,
            help_text="Indiquez le nom et le prénom, sous la forme Prenom NOM (sans aucun accent)")
    homedir = models.CharField("répertoire personnel (home)", max_length=256,
            default=HOME_BASE, unique=True,
            help_text="Laissez à <em>" + HOME_BASE + "</em> et la valeur sera calculée automatiquement")
    shell = models.CharField("interpréteur de commande (shell)", max_length=64, 
            choices=SHELLS, default=SHELLS[0][0], blank=False)
    # informations shadow (pour expire ; les autres ne sont pas utilisées pour l'instant)
    lstchg = ExpireField("dernier changement du mot de passe", default = 1,
            help_text="date du dernier changement de mot de passe")
    min = models.IntegerField("durée minimale du mot de passe", default = 0,
            help_text="nombre de jours à attendre avant de pouvoir changer le mot de passe")
    warn = models.IntegerField("durée d'avertissement", default = 0, 
            help_text="nombre de jours d'avertissement avant la fin mot de passe")
    max = models.IntegerField("durée maximale du mot de passe", default = 99999,
            help_text="nombre de jours après lesquels le mot de passe doit être changé")
    inact = models.IntegerField("", default = 0,
            help_text="nombre de jours après la fin de validité provoquant la désactivation du compte")
    expire = ExpireField("date d'expiration", default=default_expire)
    flag = models.IntegerField("réservé", default = 0)
    # champs spécifiques "auf" : source de provenance des données et dates automatiques
    source = models.CharField("source", max_length=10, default="LOCAL")
    creation = models.DateTimeField("date de création", auto_now_add=True)
    modification = models.DateTimeField("dernière modification", auto_now=True)

    def __unicode__(self):
        """Représentation d'un utilisateur : son login suivi de son uid"""
        return "%s (%s)" % (self.username, self.uid)

    class Meta:
        db_table = "users"
        verbose_name = "utilisateur système (NSS)"
        verbose_name_plural = "utilisateurs système (NSS)"
        get_latest_by = 'uid'

    def save(self, force_insert=False, force_update=False, force_source=False):
        """
        Enregistre ou modifie un utilisateur dans la table ``users``.

        La methode User.save() effectue quelques vérifications et modification
        avant de procéder à l'enregistrement réel de l'utilisateur dans la base
        de données :

         * si l'utilisateur (indiqué par son username) existe déjà, les champs
           ``uid`` et ``source`` ne sont pas modifiables (exception levée)
         * si ``source`` n'est pas 'LOCAL', alors l'enregistrement n'est pas
           effectué, sauf si on précise *force_source=True* lors de l'appel à
           *save()*
         * si aucun ``uid`` n'est imposé, alors on en calcule un nouveau, qui
           doit être supérieur à ``conf.MIN_UID`` (configurable dans le fichier 
           ``conf.py``)
         * si le ``homedir`` n'est pas précisé, on le calcule à travers une
           fonction ``homedir(self)``, qui peut être personnalisée dans le
           fichier ``conf.py``
         * si le mot de passe est livré en clair, on le crypte
        
        Ce n'est qu'après toutes ces opérations que l'enregistrement est
        effectué.
        """
        # surcharges avant enregistrement :
        # on interdit les changements sur "uid" ou "source" avec ça :
        try:
            # si l'objet existe déjà, on récupère ses données actuelles
            current = User.objects.get(pk=self.pk) 
        except:
            # si l'utilisateur n'existe pas encore : vérification du format du username
            username_ok = 'abcdefghijklmnopqrstuvwxyz0123456789-._'
            for c in self.username:
                if c not in username_ok:
                    raise ValueError("username ne doit contenir que des minuscules ASCII, des chiffres, - _ ou .")
        else:
            # on refuse de modifier l'uid ou la source, il y a trop
            # d'implications par ailleurs
            if self.source != current.source:
                raise ValueError('changement de source interdit')
            if self.uid != current.uid:
                raise ValueError("changement d'uid interdit")
        # on n'enregistre que si la source est 'LOCAL' (pour forcer
        # l'enregistrement, il faut ajouter un force_source=True lors de
        # l'appel)
        if self.source != "LOCAL" and not force_source:
            raise ValueError("source != LOCAL nécessite save(force_source=True)")
        # on vérifie le format de GECOS (ascii)
        gecos_ok = ' abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._'
        for c in self.gecos:
            if c not in gecos_ok:
                raise ValueError("GECOS ne doit contenir que des lettres ASCII, des chiffres, - _ ou .")
        # si l'objet n'a pas d'UID proposé, on en calcule un,
        # en sachant qu'il doit être plus grand que MIN_UID
        if self.uid == None:
            try:
                self.uid = max(User.objects.latest().uid + 1, MIN_UID)
            except:
                self.uid = MIN_UID
        # si le homedir n'a pas été précisé, on le calcule
        if self.homedir == HOME_BASE:
            self.homedir = homedir(self)
        # on crypte le mot de passe (si besoin)
        self.password = password_crypt( self.password )
        # et on lance l'enregistrement via la méthode mère
        return super(User, self).save(force_insert, force_update)

    def active(self):
        return self.expire >= datetime.date.today()
    active.short_description = "Actif ?"
    active.boolean = True
    active.admin_order_field = 'expire'

    def colored_expire(self):
        if self.expire == ExpireField.INFINITY:
            return u'<span style="color: green;">jamais</span>'
        delta = (self.expire - datetime.date.today()).days
        if delta < 0:
            style = "text-decoration: line-through;"
        elif delta < 8:
            style = "color: red;"
        elif delta < 15:
            style = "color: orange;"
        else:
            style = "color: green;"
        return u'<span style="%s">%s (%s)</span>' % (style, self.expire, delta)
    colored_expire.short_description = "Expire le"
    colored_expire.allow_tags = True
    colored_expire.admin_order_field = 'expire'

    def groups(self):
        """
        Retourne la liste de tous les groupes auquel appartient
        un utilisateur (groupe principal et groupes secondaires)
        """
        groups=[self.gid]
        groups.extend(self.secgroups.all())
        return groups


class Group(models.Model):
    """
    Représentation d'un groupe au sens Unix (nss):

     * ``gid`` : numéro identifiant le groupe
     * ``name`` : nom du groupe, uniquement composé de lettres minuscules et de chiffres
     * ``password`` : mot de passe du groupe, en général pas utilisé et donc positionné à 'x'
    """
    name = models.CharField("nom du groupe", max_length=32, unique=True,
            help_text="Indiquez un identifiant de groupe, uniquement composé de lettres minuscules et de chiffres")
    password = models.CharField("mot de passe (crypté)", max_length=64, default='x', editable=False)
    gid = models.AutoField("groupID", primary_key=True) 
    # utilisateurs qui ont ce groupe en secondaire
    users_secgroup = models.ManyToManyField("User", related_name='users_secgroup', through='GroupList', blank=True)

    def __unicode__(self):
        """Représentation d'un groupe : son nom suivi de son gid"""
        return "%s (%s)" % (self.name, self.gid)

    class Meta:
        db_table = "groups"
        verbose_name = "groupe système (NSS)"
        verbose_name_plural = "groupes système (NSS)"
        get_latest_by = 'gid'

    def save(self, force_insert=False, force_update=False):
        """
        Enregistre ou modifie un groupe dans la table ``groups``.

        Si aucun gid n'est proposé, la methode Group.save() en calcule un
        avant de procéder à l'enregistrement réel du groupe dans la base.
        """
        # si l'objet n'a pas de GID proposé, on en calcule un,
        # en sachant qu'il doit être plus grande que MIN_GID
        if self.gid == None:
            try:
                self.gid = max(Group.objects.latest().gid + 1, MIN_GID)
            except:
                self.gid = MIN_GID
        # on crypte le mot de passe (si besoin)
        self.password = password_crypt( self.password )
        # et on enregistre l'objet "pour de vrai"
        return super(Group, self).save(force_insert, force_update)


    def users(self):
        """
        Retourne la liste des utilisateurs compris dans ce groupe,
        que ce groupe soit leur groupe primaire ou secondaire.
        """
        users=[]
        for user in self.user_set.all():
            users.append(user)
        for user in self.users_secgroup.all():
            users.append(user)
        return users


class GroupList(models.Model):
    """
    Association entre un *User* et un *Group* afin de connaître les groupes
    secondaires auxquels appartient un utilisateur donné.
    """
    gid = models.ForeignKey(Group, db_column="gid")
    username = models.ForeignKey(User, db_column="username")

    def __unicode__(self):
        return "%s dans %s" % (self.username, self.gid)

    class Meta:
        db_table = "grouplist"
        verbose_name = "appartenance groupe secondaire"
        verbose_name_plural = "appartenances groupes secondaires"
        unique_together = ( "gid", "username" )

